iroh/src/client/blobs.rs

```diff
@@ -520,6 +554,92 @@ impl Future for BlobDownloadProgress {
     }
 }
 
+/// Outcome of a blob export operation.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct BlobExportOutcome {
+    /// The total size of the exported data.
+    total_size: u64,
+}
+
+/// Progress stream for blob export operations.
+#[derive(derive_more::Debug)]
+pub struct BlobExportProgress {
+    #[debug(skip)]
+    stream: Pin<Box<dyn Stream<Item = Result<ExportProgress>> + Send + Unpin + 'static>>,
+    current_total_size: Arc<AtomicU64>,
+}
+
+impl BlobExportProgress {
+    /// Create a `BlobExportProgress` that can help you easily poll the `ExportProgress` stream from your download until it is finished or errors.
+    pub fn new(
+        stream: (impl Stream<Item = Result<impl Into<ExportProgress>, impl Into<anyhow::Error>>>
+             + Send
+             + Unpin
+             + 'static),
+    ) -> Self {
+        let current_total_size = Arc::new(AtomicU64::new(0));
+        let total_size = current_total_size.clone();
+        let stream = stream.map(move |item| match item {
+            Ok(item) => {
+                let item = item.into();
+                if let ExportProgress::Found { size, .. } = &item {
+                    let size = size.value();
+                    total_size.fetch_add(size, Ordering::Relaxed);
+                }
+
+                Ok(item)
+            }
+            Err(err) => Err(err.into()),
+        });
+        Self {
+            stream: Box::pin(stream),
+            current_total_size,
+        }
+    }
+
+    /// Finish writing the stream, ignoring all intermediate progress events.
+    ///
+    /// Returns a [`BlobExportOutcome`] which contains the size of the content we exported.
+    pub async fn finish(self) -> Result<BlobExportOutcome> {
+        self.await
+    }
+}
+
+impl Stream for BlobExportProgress {
+    type Item = Result<ExportProgress>;
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.stream.poll_next_unpin(cx)
+    }
+}
+
+impl Future for BlobExportProgress {
+    type Output = Result<BlobExportOutcome>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        loop {
+            match self.stream.poll_next_unpin(cx) {
```

Comment:
I like the style of this API, but I don't like the manual poll impl. Any way to get the same thing without having to manually write poll, or at least reuse this somehow? I don't want our entire codebase to be full of this stuff. Writing new RPC methods is boilerplatey enough as it is.

Reply:
We can also postpone this and make an issue.

Reply:
Yes let's move this to a followup. This applies to all these progress streams. I'm quite confident we can find an abstraction to reduce the boilerplate here.

Created an issue: https://github.com/n0-computer/iroh/issues/2114